internal/mux/buffer_test.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2020 The Oto Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mux_test
+
+import (
+	"io"
+	"sync"
+	"testing"
+
+	"github.com/hajimehoshi/oto/internal/mux"
+)
+
+const concurrency = 1000
+
+func TestBufferConcurrentWrites(t *testing.T) {
+	b := &mux.ConcurrentBuffer{}
+
+	ch := make(chan struct{})
+	var wg sync.WaitGroup
+	wg.Add(concurrency)
+	for i := 0; i < concurrency; i++ {
+		go func() {
+			<-ch
+			b.Write([]byte{1})
+			wg.Done()
+		}()
+	}
+	close(ch)
```

Comment:
Is it OK to close the channel here?

Reply:
```
	var wg, wgStart sync.WaitGroup
	wgStart.Add(1)
	wg.Add(concurrency)
	for i := 0; i < concurrency; i++ {
		go func() {
			wgStart.Wait()
			b.Write([]byte{1})
			wg.Done()
		}()
	}
	wgStart.Done()
```

Reply:
Its basically equivalent to that. The intent is to wake all goroutines at the same time to maximize chance of collisions.

Reply:
I'll swap the code to this to make it clear what's going on.